drivers/video/arducam_mega.c

```diff
@@ -0,0 +1,1048 @@
+				   uint32_t img_len, uint8_t first)
+{
+	uint8_t cmd_fifo_read[] = {BURST_FIFO_READ, 0x00};
+	uint8_t buf_len = first == 0 ? 1 : 2;
+
+	struct spi_buf tx_buf[] = {
+		{.buf = cmd_fifo_read, .len = buf_len},
+	};
+
+	struct spi_buf_set tx_bufs = {.buffers = tx_buf, .count = 1};
+
+	struct spi_buf rx_buf[2] = {
+		{.buf = cmd_fifo_read, .len = buf_len},
+		{.buf = img_buff, .len = img_len},
+	};
+	struct spi_buf_set rx_bufs = {.buffers = rx_buf, .count = 2};
+
+	return spi_transceive_dt(spec, &tx_bufs, &rx_bufs);
+}
+
+static int arducam_mega_await_bus_idle(const struct spi_dt_spec *spec, uint8_t tries)
+{
+	while ((arducam_mega_read_reg(spec, CAM_REG_SENSOR_STATE) & 0x03) != SENSOR_STATE_IDLE) {
+		if (tries-- == 0) {
+			return -1;
+		}
+		k_msleep(2);
+	}
+
+	return 0;
+}
+
+static int arducam_mega_set_brightness(const struct device *dev, enum MEGA_BRIGHTNESS_LEVEL level)
+{
+	int ret = 0;
+	const struct arducam_mega_config *cfg = dev->config;
+
+	ret |= arducam_mega_await_bus_idle(&cfg->bus, 3);
+
+	ret |= arducam_mega_write_reg(&cfg->bus, CAM_REG_BRIGHTNESS_CONTROL, level);
+
+	if (ret == -1) {
+		LOG_ERR("Failed to set brightness level %d", level);
+	}
+
+	return ret;
+}
+
+static int arducam_mega_set_saturation(const struct device *dev, enum MEGA_SATURATION_LEVEL level)
+{
+	int ret = 0;
+	const struct arducam_mega_config *cfg = dev->config;
+
+	ret |= arducam_mega_await_bus_idle(&cfg->bus, 3);
+
+	ret |= arducam_mega_write_reg(&cfg->bus, CAM_REG_SATURATION_CONTROL, level);
+
+	if (ret == -1) {
+		LOG_ERR("Failed to set saturation level %d", level);
+	}
+
+	return ret;
+}
+
+static int arducam_mega_set_contrast(const struct device *dev, enum MEGA_CONTRAST_LEVEL level)
+{
+	int ret = 0;
+	const struct arducam_mega_config *cfg = dev->config;
+
+	ret |= arducam_mega_await_bus_idle(&cfg->bus, 3);
+
+	ret |= arducam_mega_write_reg(&cfg->bus, CAM_REG_CONTRAST_CONTROL, level);
+
+	if (ret == -1) {
+		LOG_ERR("Failed to set contrast level %d", level);
+	}
+
+	return ret;
+}
+
+static int arducam_mega_set_EV(const struct device *dev, enum MEGA_EV_LEVEL level)
+{
+	int ret = 0;
+	const struct arducam_mega_config *cfg = dev->config;
+
+	ret |= arducam_mega_await_bus_idle(&cfg->bus, 3);
+
+	ret |= arducam_mega_write_reg(&cfg->bus, CAM_REG_EV_CONTROL, level);
+
+	if (ret == -1) {
+		LOG_ERR("Failed to set contrast level %d", level);
+	}
+
+	return ret;
+}
+
+static int arducam_mega_set_sharpness(const struct device *dev, enum MEGA_SHARPNESS_LEVEL level)
+{
+	int ret = 0;
+	struct arducam_mega_data *drv_data = dev->data;
+	struct arducam_mega_info *drv_info = drv_data->info;
+	const struct arducam_mega_config *cfg = dev->config;
+
+	if (!drv_info->enable_sharpness) {
+		LOG_ERR("This device does not support set sharpness.");
+		return -1;
+	}
+
+	ret |= arducam_mega_await_bus_idle(&cfg->bus, 3);
+
+	ret |= arducam_mega_write_reg(&cfg->bus, CAM_REG_SHARPNESS_CONTROL, level);
+
+	if (ret == -1) {
```

Comment:
Given the `-errno` being returned for errors, I suppose this would need to be converted to `ret != 0` or `ret < 0` (to support cases when ret > 0 to indicate status).